### PR TITLE
PoC implementation of Waku Pairing and Secure Transfer

### PR DESCRIPTION
**Pull request by s1fr0**
This PR addresses the third item of https://github.com/status-im/nwaku/issues/1065 by implementing the construction proposed in the currently under review RFC https://github.com/vacp2p/rfc/pull/531.

This PR will be updated with any eventual feedback received in https://github.com/vacp2p/rfc/pull/531.

~~In contrast to what described in the RFC, this PR **does not** implement at the moment the message identification technique based on `messageNametag` and detailed in https://github.com/status-im/nwaku/issues/1081#issue-1342350736 since this will require an update to [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/) RFC and a minor refactor of the Noise module: this will most probably be addressed by a follow-up PR.~~

EDIT: Full message identification nametag logic implemented.

### CODE REVIEW

**Comment by LNSD** [comment]
Can you split the tests into different test files? The test file is becoming too big 😅

**Review by LNSD — APPROVED** [review]
LGTM (x2) ✅😁

File: waku/v2/protocol/waku_noise/noise_utils.nim [hunk 1]
```diff
@@ -57,6 +57,76 @@ proc pkcs7_unpad*(payload: seq[byte], paddingSize: int): seq[byte] =
...
+# Converts a sequence or array (arbitrary size) to a MessageNametag
+proc toMessageNametag*(input: openArray[byte]): MessageNametag =
+  var byte_seq: seq[byte] = @input
+
+  # We set its length to the default message nametag length (will be truncated or 0-padded)
+  byte_seq.setLen(MessageNametagLength)
+  
+  # We copy it to a MessageNametag
+  var messageNametag: MessageNametag
+  for i in 0..<MessageNametagLength:
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise_utils.nim:113]:
```suggestion
  for i in 0.. < MessageNametagLength:
```

File: waku/v2/protocol/waku_noise/noise_utils.nim [hunk 2]
```diff
@@ -57,6 +57,76 @@ proc pkcs7_unpad*(payload: seq[byte], paddingSize: int): seq[byte] =
...
+  let committedStaticKey = seqToDigest256(decode(values[4]).toBytes)
+ 
+  return (applicationName, applicationVersion, shardId, ephemeralKey, committedStaticKey)
+
+# Converts a sequence or array (arbitrary size) to a MessageNametag
+proc toMessageNametag*(input: openArray[byte]): MessageNametag =
+  var byte_seq: seq[byte] = @input
+
+  # We set its length to the default message nametag length (will be truncated or 0-padded)
+  byte_seq.setLen(MessageNametagLength)
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise_utils.nim:109]: Wouldn't it be simpler to loop through `input` and put `MessageNametagLength` many bytes into `messageNametag`? I mean  lines 106 and 109 can be skipped

File: waku/v2/protocol/waku_noise/noise_utils.nim [hunk 3]
```diff
@@ -138,6 +208,85 @@ proc hashProtocol*(protocolName: string): MDigest[256] =
...
+  hashInput.add getBytes(publicKey)
+  hashInput.add r
+
+  # The output hash value
+  var hash: MDigest[256]
+  hash = sha256.digest(hashInput)
+
+  return hash
+
+proc genAuthcode*(hs: HandshakeState): string =
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise_utils.nim:224]: Worth a unittest

File: waku/v2/protocol/waku_noise/noise_utils.nim [hunk 4]
```diff
@@ -138,6 +208,85 @@ proc hashProtocol*(protocolName: string): MDigest[256] =
...
+  var output: array[1, array[8, byte]]
+  sha256.hkdf(hs.ss.h.data, [], [], output)
+  let code = cast[uint64](output[0]) mod 100_000_000
+  return $code
+
+proc initNametagsBuffer*(mntb: var MessageNametagBuffer) =
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise_utils.nim:230]: That would be great if we could have an explanation of the name tag usage somewhere in RFC or in here.  Without proper description, it is hard to judge the implementation.

**Review by staheri14 — COMMENTED** [review]
Looks good to me! left some comments.

**Pull request merged**